### PR TITLE
[Color][Service Layer][MWPW-187667] Stock API Integration

### PR DESCRIPTION
**Pull request by vvineett**
## Summary

Adds a new **Stock plugin** to the service layer, enabling Adobe Stock integration for color-based image searches, curated gallery lookups, data availability checks, and redirect URL generation. The plugin follows the same `BaseApiService` architecture as existing plugins (Kuler, Behance) and includes a `StockProvider` wrapper for ergonomic consumption. Stock API responses are normalized by mapping `files` to `themes` for cross-plugin consistency.

---

## Jira Ticket

Resolves: [MWPW-187667](https://jira.corp.adobe.com/browse/MWPW-187667)

---

## Test URLs

| Env | URL |
|-------------|-----|
| **Before**  | https://main--da-express-milo--adobecom.aem.page/express/ |
| **After**   | https://vvineett-color-service-layer-stock--da-express-milo--adobecom.aem.page/drafts/dhananjay/color-poc/?martech=off |

---

### New Files

| File | Purpose |
|------|---------|
| `libs/services/plugins/stock/StockPlugin.js` | Main plugin class extending `BaseApiService`. Registers 4 action groups and injects custom `x-product: AdobeColor/4.0` header. |
| `libs/services/plugins/stock/actions/StockActions.js` | Four action group classes: `SearchActions`, `GalleryActions`, `DataActions`, `RedirectActions`. |
| `libs/services/plugins/stock/topics.js` | Topic constants (`stock.search.files`, `stock.gallery.*`, `stock.data.*`, `stock.redirect.*`) and action group name map. |
| `libs/services/plugins/stock/constants.js` | Curated gallery list (`Wilderness`, `Flavour`, `Travel`) and default batch size (`36`). |
| `libs/services/plugins/stock/index.js` | Plugin manifest with lazy-loading and `ENABLE_STOCK` feature flag. |
| `libs/services/plugins/stock/README.md` | Full plugin documentation with usage examples and response schemas. |
| `libs/services/providers/StockProvider.js` | Provider wrapper with safe-execution methods for all Stock operations. |

### Modified Files

| File | Change |
|------|--------|
| `libs/services/plugins/index.js` | Registered Stock plugin manifest alongside existing plugins. |
| `libs/services/providers/index.js` | Exported `StockProvider` and `createStockProvider` factory. |
| `libs/services/config.js` | Added Stock service config for both prod (`stock.adobe.io`) and stage (`stock-stage.adobe.io`) environments. |

### Action Groups & Topics

| Action Group | Topic | Description |
|-------------|-------|-------------|
| **Search** | `stock.search.files` | Queries Stock Search/Files API with pagination (offset-based, 36 items/page). Filters: `content_type:photo`, `premium: false`, locale `en-US`. |
| **Gallery** | `stock.gallery.getCuratedList` | Returns hardcoded curated gallery names (`Wilderness`, `Flavour`, `Travel`). |
| **Gallery** | `stock.gallery.getByName` | Delegates to search using gallery name as query; returns `undefined` for non-curated names. |
| **Data** | `stock.data.checkAvailability` | Uses raw `fetch()` with plugin headers to check if an endpoint returns data. |
| **Redirect** | `stock.redirect.getFileUrl` | Builds `https://stock.adobe.com/images/id/{fileId}` (sync, no HTTP). |
| **Redirect** | `stock.redirect.getContributorUrl` | Builds `https://stock.adobe.com/contributor/{creatorId}` (sync, no HTTP). |

### Provider API (`StockProvider`)

| Method | Returns | Description |
|--------|---------|-------------|
| `searchThemes(query, { page })` | `{ themes, nb_results, hasMore }` or `null` | Search Stock images with pagination |
| `getCuratedGalleries()` | `{ themes: [{ title }] }` or `null` | Get curated gallery names |
| `getGalleryByName(name, { page })` | Search results or `null`/`undefined` | Fetch gallery by curated name |
| `checkDataAvailability(endpoint)` | `boolean` or `null` | Check if endpoint has data |
| `getFileRedirectUrl(fileId)` | `string` or `null` | Build Stock file URL (sync) |
| `getContributorUrl(creatorId)` | `string` or `null` | Build contributor URL (sync) |

### Key Design Decisions

- **Feature flag** `ENABLE_STOCK` defaults to enabled (must be explicitly set to `false` to disable).
- **Response normalization**: Stock `files` array is renamed to `themes` and `hasMore` is computed from `offset + themes.length < nb_results` for consistency with Kuler/Behance plugins.
- **No auth required**: Uses API key (`ColorWeb`) only; no user login token needed.
- **Safe execution**: All provider methods wrap calls in `safeExecute`/`safeExecuteSync`, returning `null` on failure with error logging via `window.lana`.

---

## Verification Steps

1. Navigate to the **After** URL above.
2. Open DevTools Network tab and filter for `stock.adobe.io`.
3. Verify that Stock search requests include:
   - `x-api-key: ColorWeb` header
   - `x-product: AdobeColor/4.0` header
   - Proper `search_parameters[words]`, `content_type:photo`, `premium: false` query params
4. Confirm search results return with `themes` array and pagination (`hasMore`, `nb_results`).
5. Verify curated galleries return the expected names (`Wilderness`, `Flavour`, `Travel`).
6. Confirm redirect URLs resolve correctly (e.g., `stock.adobe.com/images/id/{id}`).

---

## Potential Regressions

- [https://vvineett-color-service-layer-stock--da-express-milo--adobecom.aem.page/drafts/dhananjay/color-poc/?martech=off](https://vvineett-color-service-layer-stock--da-express-milo--adobecom.aem.page/drafts/dhananjay/color-poc/?martech=off)
- Existing plugins (Kuler, Behance, CC Library, Curated, Universal) should remain unaffected — Stock is a net-new plugin with its own manifest entry and lazy-loaded independently.

---

## Additional Notes

- The Stock plugin is not yet consumed by any block. The `color-explorer-hybrid` block still uses mock data. Future PRs will wire up Stock data into the block UI.
- Test files added under `libs/services/test/services/stock/` covering plugin activation, action registration, search, gallery, data, and redirect logic.
- Stage environment routes to `stock-stage.adobe.io` / `primary.stock.stage.adobe.com`.

### CODE REVIEW

File: express/code/libs/services/plugins/stock/StockPlugin.js [hunk 1]
```diff
@@ -1,11 +1,19 @@
 import BaseApiService from '../../core/BaseApiService.js';
-import StockActions from './actions/StockActions.js';
 import { StockActionGroups } from './topics.js';
+import SearchActions from './actions/SearchActions.js';
```
Comment (fullcolorcoder) [inline express/code/libs/services/plugins/stock/StockPlugin.js]: This is a nice separation of concerns, but FYI this increases module requests (1 actions file → 4). Since Franklin has no bundler, extra modules can create waterfall. Consider consolidating the 4 action groups into a single actions/StockActions.js that exports the four classes (still <300 lines) to reduce network overhead while keeping logical separation.

IE: 

```
import {
  SearchActions,
  GalleryActions,
  DataActions,
  RedirectActions,
} from './actions/StockActions.js';
```

@yeibercano thoughts here? This still will pile up fast, and we need to see the waterfall impacts in these PRs or it could be messier to track down later.

**Review by fullcolorcoder — CHANGES_REQUESTED** [review]
A few comments here, and the same concerns / asks as the other PRs

- Unit tests
- Testing instructions & Demonstration of the action 
- Before & After links both linking to a block that shows this in action to measure performance

File: express/code/libs/services/plugins/stock/StockPlugin.js (hunk 1 above)
Comment (cloakedj) [reply express/code/libs/services/plugins/stock/StockPlugin.js]: @fullcolorcoder That is a very valid point. There are 2 things we can do in my opinion - 
1. We can move all actions into a single actions.js but keep their implementations separate still, for clarity. This would reduce the network request depth. The only issue is actions.js will be a bigger file.
2. A second more advanced solution could be importing actions via dispatch. When a dispatch would take place for an action, we would see if it is loaded or not, and only load it then. Our plugins are already only loaded on demand. This will still allow us to keep our action grouping separate, but there is the additional concern around network kicking off before the topic dispatch or similar edge cases that could cause the action to not download and hence dispatch would fail. 

Which approach do you think we should go with?
cc: @vvineett

Comment (yeibercano) [reply express/code/libs/services/plugins/stock/StockPlugin.js]: @fullcolorcoder, yes, I agree overall with the concerns. We did talk about this night time standup. @cloakedj  is working on it: long chains of import is going to be a battle for this kind of project. We have to minimize it.

File: express/code/libs/services/plugins/stock/topics.js [hunk 2]
```diff
@@ -2,19 +2,33 @@
  * Stock Plugin Topics
  *
  * Topics define the available actions for the Stock service.
+ * Topic names are generic and functionality-focused.
+ * Use these with plugin.dispatch() for direct API access,
+ * or use the StockProvider for a friendlier interface.
  */
 export const StockTopics = {
```
Comment (yeibercano) [inline express/code/libs/services/plugins/stock/topics.js:1]: Small: plain object  camelCase > `stockTopics` particularly since you are using `class` around this API.

File: express/code/libs/services/plugins/stock/topics.js [hunk 3]
```diff
@@ -2,19 +2,33 @@
...
+  REDIRECT: {
+    GET_FILE_URL: 'stock.redirect.getFileUrl',
+    GET_CONTRIBUTOR_URL: 'stock.redirect.getContributorUrl',
   },
 };
 
 /**
  * Action group identifiers
  */
 export const StockActionGroups = {
```
Comment (yeibercano) [inline express/code/libs/services/plugins/stock/topics.js:18]: camelCase `stockActionGroups`

File: express/code/libs/services/providers/StockProvider.js [hunk 4]
```diff
@@ -60,53 +57,72 @@ export default class StockProvider extends BaseProvider {
...
+  /**
+   * Get gallery by name (searches Stock with gallery name as query)
+   *
+   * @param {string} name - Gallery name
+   * @param {Object} [options] - Search options
+   * @returns {Promise<Object|null|undefined>} Gallery data or null/undefined on failure or unknown name
+   */
+  async getGalleryByName(name, options = {}) {
+    const criteria = this.#transformSearchParams(name, options);
+    return this.safeExecute(() => this.#actions.getGalleryByName(criteria));
```
Comment (yeibercano) [inline express/code/libs/services/providers/StockProvider.js:69]: [ @returns {Promise<Object|null|undefined>} Gallery data or null/undefined on failure or unknown name](https://github.com/adobecom/da-express-milo/pull/125/changes#diff-154ab2981b83656c0b91d2a994271f409e4813e17c44c4e04646a71ee25c2ac6R86) says you are returning a Promise: can you clarify and match?

File: test/services/plugins/stock/StockPlugin.test.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+  describe('serviceName', () => {
+    it('should return "Stock"', () => {
+      expect(StockPlugin.serviceName).to.equal('Stock');
+    });
+  });
+
+  describe('isActivated (feature flag)', () => {
+    [
+      { label: 'ENABLE_STOCK not set', config: { features: {} }, expected: true },
```
Comment (yeibercano) [inline test/services/plugins/stock/StockPlugin.test.js:38]: Where can we find information about how to configure / enable these features?

**Comment by yeibercano** [comment]
@cloakedj overall a solid start: first review with minor comments. I second @fullcolorcoder points. One are I'd like us to cover more would be: tests for validation errors, availability edge cases, and provider error paths. 

QQ: is there any documentation to these APIs? we should create at least wiki.

File: express/code/libs/services/providers/StockProvider.js (hunk 4 above)
Comment (cloakedj) [reply express/code/libs/services/providers/StockProvider.js:69]: safeExecute returns either the object from the API, if the request was successful, or Promise resolving to null/undefined if the request fails, while logging the failure. This makes sure the block that uses this does not break and can silently handle issues. Hence it always returns a promise, so this is correct.

File: express/code/libs/services/plugins/stock/StockPlugin.js (hunk 1 above)
Comment (cloakedj) [reply express/code/libs/services/plugins/stock/StockPlugin.js]: I have moved individual actions into single StockActions.js. We will follow this similar pattern for all other plugins as well. I have updated related documentation and Agents.md as well to reflect this change as well as updated the testing framework to reflect it.

**Comment by cloakedj** [comment]
> @cloakedj overall a solid start: first review with minor comments. I second @fullcolorcoder points. One are I'd like us to cover more would be: tests for validation errors, availability edge cases, and provider error paths.
> 
> QQ: is there any documentation to these APIs? we should create at least wiki.

https://wiki.corp.adobe.com/display/adobedotcom/Color+Explorer+Service+Layer+-+API+Documentation - @yeibercano Here is the documentation. I had created this a while back. It is almost up to date. Some final tweaks are required but it has all the documentation on the entire Plugin system.

File: test/services/plugins/stock/StockPlugin.test.js (hunk 5 above)
Comment (cloakedj) [reply test/services/plugins/stock/StockPlugin.test.js:38]: https://github.com/adobecom/da-express-milo/pull/125#issuecomment-3883741536 - Here is the documentation on the same.
